File: core/src/main/java/com/gmail/marcosav2010/communicator/packet/handling/PacketAction.java
package com.gmail.marcosav2010.communicator.packet.handling;

import com.gmail.marcosav2010.communicator.packet.Packet;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;

@RequiredArgsConstructor
public class PacketAction {

    private final Runnable action, onTimeOut;
    @Getter(AccessLevel.PACKAGE)
    @Setter(AccessLevel.PACKAGE)
    private Class<? extends Packet> type;

    public void onReceive() {
        if (action != null)
            action.run();
    }

    public void onTimeOut() {
        if (onTimeOut != null)
            onTimeOut.run();
    }
}

File: core/src/main/java/com/gmail/marcosav2010/communicator/packet/handling/PacketAction.kt
package com.gmail.marcosav2010.communicator.packet.handling

import com.gmail.marcosav2010.communicator.packet.Packet

class PacketAction(private val action: () -> Unit,
                   private val onTimeOut: (() -> Unit)?,
                   internal var type: Class<out Packet?>
) {

    fun onReceive() = action.invoke()

    fun onTimeOut() = onTimeOut?.invoke()
}